File: shared/src/commonMain/kotlin/com/fedetorresdev/iosmodularizationdemo/issues/IssueDetailRepository.kt
package com.fedetorresdev.iosmodularizationdemo.issues

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

interface IssueDetailRepository {
    fun getIssueDetail(id: String): Flow<IssueDetailModel?>
}

class IssueDetailRepositoryImpl : IssueDetailRepository {
    override fun getIssueDetail(id: String): Flow<IssueDetailModel?> = flow {
        val issue = mockIssueDetails.find { it.id == id }
        emit(issue)
    }

    private val mockIssueDetails = listOf(
        IssueDetailModel(
            id = "1",
            title = "Issue 1",
            description = "Critical bug in the authentication system causing login failures for certain users",
            status = "Open",
            assignee = "John Doe",
            priority = "High",
            createdDate = "2024-10-01",
            dueDate = "2024-10-15",
            reporter = "Jane Smith"
        ),
        IssueDetailModel(
            id = "2",
            title = "Issue 2",
            description = "Feature request to add dark mode support across the entire application",
            status = "Closed",
            assignee = "Alice Johnson",
            priority = "Medium",
            createdDate = "2024-09-15",
            dueDate = "2024-10-10",
            reporter = "Bob Wilson"
        ),
        IssueDetailModel(
            id = "3",
            title = "Issue 3",
            description = "Performance optimization needed for data loading on the dashboard page",
            status = "In Progress",
            assignee = "Mike Brown",
            priority = "High",
            createdDate = "2024-10-05",
            dueDate = "2024-10-20",
            reporter = "Sarah Davis"
        )
    )
}

File: modularizediOSApp/Modules/Issues/Sources/Issues/Dependencies/IssueDetailRepository.swift
//
//  IssueDetailRepository.swift
//  Issues
//
//  Created by Federico Torres on 13/10/25.
//

import ComposableArchitecture
import CoreInterfaces
import Foundation

public struct IssueDetailRepository: Sendable {
    public var getIssueDetail: @Sendable (String) -> AsyncStream<IssueDetailUIModel?>
    
    public init(getIssueDetail: @escaping @Sendable (String) -> AsyncStream<IssueDetailUIModel?>) {
        self.getIssueDetail = getIssueDetail
    }
}

extension IssueDetailRepository: TestDependencyKey {
    public static var testValue: IssueDetailRepository {
        return Self(
            getIssueDetail: { id in
                return AsyncStream { continuation in
                    continuation.yield(
                        IssueDetailUIModel(
                            id: id,
                            title: "Test Issue",
                            description: "Test description",
                            status: "Open",
                            assignee: "Test User",
                            priority: "High",
                            createdDate: "2024-01-01",
                            dueDate: "2024-01-15",
                            reporter: "Test Reporter"
                        )
                    )
                    continuation.finish()
                }
            }
        )
    }
}

public extension DependencyValues {
    var issueDetailRepository: IssueDetailRepository {
        get { self[IssueDetailRepository.self] }
        set { self[IssueDetailRepository.self] = newValue }
    }
}

